dbt_meshify/mesh.py

```diff
@@ -0,0 +1,134 @@
+from dbt_meshify.dbt_projects import BaseDbtProject
+
+
+class ProjectDependencyType(str, Enum):
+    """ProjectDependencyTypes define how the dependency relationship was defined."""
+
+    Source = "source"
+    Package = "package"
+
+
+@dataclass
+class ProjectDependency:
+    """ProjectDependencies define shared resources between two different projects"""
+
+    upstream: str
+    downstream: str
+    type: ProjectDependencyType
+
+    def __key(self):
+        return self.upstream, self.downstream, self.type
+
+    def __hash__(self):
+        return hash(self.__key())
+
+    def __eq__(self, other):
+        if isinstance(other, ProjectDependency):
+            return self.__key() == other.__key()
+        return NotImplemented
+
+
+class Mesh:
+    """
+    Mesh computes the metagraph between separate DbtProjects. This includes
+    providing an interface for mapping the dependencies between projects be
+    it source-hacked models, imported packages, or explicit cross-project
+    references.
+    """
+
+    @staticmethod
+    def _find_relation_dependencies(
+        source_relations: Set[str], target_relations: Set[str]
+    ) -> Set[str]:
+        """
+        Identify dependencies between projects using shared relations.
+        """
+        return source_relations.intersection(target_relations)
+
+    def _source_dependencies(
+        self, project: BaseDbtProject, other_project: BaseDbtProject
+    ) -> Set[ProjectDependency]:
+        """
+        Identify source-hack dependencies between projects.
+
+        Source hacking occurs when Project A defines a model, and a downstream project (Project B)
+        defines a source for the materialization of that same model.
+        """
+
+        relations = self._find_relation_dependencies(
+            source_relations={model.relation_name for model in project.models().values()},
+            target_relations={source.relation_name for source in other_project.sources().values()},
+        )
+
+        forward_dependencies = {
+            ProjectDependency(
+                upstream=project.model_relation_names.get(relation),
+                downstream=other_project.source_relation_names.get(relation),
+                type=ProjectDependencyType.Source,
+            )
+            for relation in relations
+        }
+
+        backwards_relations = self._find_relation_dependencies(
+            source_relations={model.relation_name for model in other_project.models().values()},
+            target_relations={source.relation_name for source in project.sources().values()},
+        )
+
+        backward_dependencies = {
+            ProjectDependency(
+                upstream=other_project.model_relation_names.get(relation),
+                downstream=project.source_relation_names.get(relation),
+                type=ProjectDependencyType.Source,
+            )
+            for relation in backwards_relations
+        }
+
+        return forward_dependencies | backward_dependencies
```

Comment:
i think this pipe syntax might be problematic for lower python versions?

Reply:
The `set1 | set2` notation is supported for unions starting in Python 3.5, which is currently EOL. We should be good!

https://docs.python.org/3.5/library/stdtypes.html#frozenset.union

Reply:
beautiful! i thought that was an `or` !